### PR TITLE
Nbands convergence check (vasp)

### PR DESCRIPTION
**Pull request by sravankumarkb**
Hi all,

Here is a function I would like to have in vasp base class. This function allows me to check if the calculation is converged with respect to the nbands. It would be great if you could let me know your feedback before merging into the master. Thanks.

### CODE REVIEW

File: pyiron/vasp/base.py [hunk 1]
```diff
@@ -400,6 +400,12 @@ def collect_output(self):
                 exclude_nodes=self._exclude_nodes_hdf,
             )
 
+    def nbands_convergence_check(self):
+        if (max(self["output/electronic_structure/occ_matrix"][:,-1])>0):
```
Comment (samwaseda) [inline pyiron/vasp/base.py:404]: I see only two indices but shouldn't there be three? I presume the first is k-points and the second is bands. Where's spin?

Comment (samwaseda) [reply pyiron/vasp/base.py:404]: And I guess we can make it shorter:

```
def nbands_convergence_check(self):
    return np.all(np.isclose(self["output/electronic_structure/occ_matrix"][:,-1], 0))
```

Comment (samwaseda) [reply pyiron/vasp/base.py:404]: Oh wait there should be even four indices since there should be also ionic steps.

Comment (sravankumarkb) [reply pyiron/vasp/base.py]: Thanks Sam for the comments, I ran a couple of calculations with spin and more than one ionic steps, the shape of the occupancy matrix does not change and remains the same as (kpoints, nbands).  I have made the function shorter as you suggested and added a docstring. 
If this satisfies your comments, you can resolve the conversation. Thanks.

Comment (samwaseda) [reply pyiron/vasp/base.py]: Ok I created a new issue and hopefully @sudarsan-surendralal will reply soon. Feel free to merge it now or after Sudarsan corrects the bug.

Comment (sudarsan-surendralal) [reply pyiron/vasp/base.py]: The shape in `output/electronic_structure` should be `n_spin x n_kpoints x n_bands`. This is the electronic structure of the final step only. The spin not being there is a bug!

Comment (samwaseda) [reply pyiron/vasp/base.py]: I agree that the most interesting one is the final occupancy, but shouldn't we also check the occupancies of all steps? This being said, I admit it's possible that the amount of data to store might become too large if there are too many steps...

**Comment by sudarsan-surendralal** [comment]
@sravankumarkb Can you now merge the master branch into this branch and try Sam's suggestions?

**Comment by sravankumarkb** [comment]
> @sravankumarkb Can you now merge the master branch into this branch and try Sam's suggestions?

Thank you @sudarsan-surendralal for updating the electronic structure shape in the output to include the spin. I have merged the master branch and updated the nbands_convergence_check function to reflect the suggestions of @samwaseda 

Now the nbands_convergence_check function works for both spin polarized and non spin polarized calcs. 
If you guys are happy with the current function, maybe we can consider merging it into the master? 
Thanks.

File: pyiron/vasp/base.py [hunk 2]
```diff
@@ -403,6 +403,16 @@ def collect_output(self):
...
+        """
+        Function to check the convergence of nbands in the Vasp calculation
+
+        Returns:
+
+            bool : True if the highest band is unoccupied, False if the highest band is occupied
+        """
+        return np.all(np.isclose(self["output/electronic_structure/occ_matrix"][:,:,-1], 0)) #shape is n_spin x n_kpoints x n_bands
+
     def convergence_check(self):
```
Comment (sudarsan-surendralal) [inline pyiron/vasp/base.py:416]: Nice implementation. Could you also call `nbands_convergence_check` inside `convergence_check` and then set the status to `not_converged` if the bands aren't converged?

Comment (sravankumarkb) [reply pyiron/vasp/base.py:416]: Thanks @sudarsan-surendralal  for the suggestion. I will start working on it

Comment (sudarsan-surendralal) [reply pyiron/vasp/base.py:416]: Actually, can you move this function to `GenericDFTJob`? Since this is useful for both VASP and SPHINX

**Review by jan-janssen — CHANGES_REQUESTED** [review]
Please add a unit test and currently the function is not used by default is this intended? Or should the function be used to check the result is converged? Meaning if it returns true the status should be finished and if it returns false the status should be not converged?

**Comment by samwaseda** [comment]
I added a test

**Comment by sravankumarkb** [comment]
> I added a test

Thank you @samwaseda :)

**Comment by sravankumarkb** [comment]
> Please add a unit test and currently the function is not used by default is this intended? Or should the function be used to check the result is converged? Meaning if it returns true the status should be finished and if it returns false the status should be not converged?

Thanks @jan-janssen for the suggestion. I have not thought about making the function default and therefore setting the job status, could you please let me know what I need to add to make it default? Thanks.

**Comment by samwaseda** [comment]
> Thanks @jan-janssen for the suggestion. I have not thought about making the function default and therefore setting the job status, could you please let me know what I need to add to make it default? Thanks.

Since we don't have a proper error/warning handling, I would say it's okay for now that this function is not called automatically for now. In addition, I think it's too severe if we consider it as not converged only because there's no empty band left.

**Comment by sravankumarkb** [comment]
> > Thanks @jan-janssen for the suggestion. I have not thought about making the function default and therefore setting the job status, could you please let me know what I need to add to make it default? Thanks.
> 
> Since we don't have a proper error/warning handling, I would say it's okay for now that this function is not called automatically for now. In addition, I think it's too severe if we consider it as not converged only because there's no empty band left.

Thanks @samwaseda for the clarification. Does that mean that we can merge this or is there anything else I need to do to get this merged? Thanks.

**Review by samwaseda — APPROVED** [review]

**Comment by jan-janssen** [comment]
Please transfer the pull request to https://github.com/pyiron/pyiron_atomistics

**Pull request closed without merging**